Some changes on JFish for J!1.7

quodo opened the pull request:
Hi Alex,

See "Developers version on Github" topic in "Joom!Fish 2.6 for Joomla 1.7+" board (http://www.joomfish.net/forum/viewtopic.php?f=48&t=10770) for some history. Alex wrote there 
"If we get a pull request from you in the end this is perfect".

I just could not resist... learn git and get you an easy way of reviewing the changes mentioned in my post.

In short these are the changes I made after I forked your jf-future (https://github.com/quodo/jf-future):
^ moved component language files for admin --> xml description gets translated during install when language files are in admin/language/en-GB/....sys.ini, moved other language files there as well (perhaps .menu.ini can be removed?)
^ changed manifest to reflect that move
^ changed mosPageNav to JPagination (2x) --> this was legacy (J!1.0) pagination 
^ ElementsController task now protected instead of private --> not sure about this (!), should more vars be protected instead of private since the parent class holds protected vars and not private ones

With these changes the main errors are gone, but I cannot translate anything and JoomFish thinks everything is translated. See http://www.joomfish.net/forum/viewtopic.php?f=48&t=10770#p36964 on that. I think it has something to do with id's that are available when they shouldn't be (e.g. no translation in database table). It works differently in this 3.0.0alpha than in the 2 version that I have on my computer (checked with Eclipse). I'd love to test further when that's fixed though.

With kind regards,
Mirjam (username quodo)

akempkens commented:
Now let's move on with learning github :-). I will try to verify and check the different changes. 
One thing git is already telling me: this pull request can be automatically merged - sounds like a good thing and is green.

akempkens commented on component/admin/language/en-GB/en-GB.com_joomfish.ini:
```diff
@@ -1,412 +1,412 @@
```
I do not understand why git removes and adds this whole file. As far as I see it you primarily moved it around.

quodo commented:
Correct: I just moved the file and I also don't understand why git does this.

Hmmm, I noticed that when I don't change a file, but do save it (e.g. add a comment sign, save the file, remove the comment sign, save the file: result is the same file with a different date) TortoiseGit shows the file as being changed (as opposed to TortoiseSVN: that only checks the contents of the file). Could it be such a thing? I don't know how to check that.

Next time I'll just use (rightclick) TortoiseGit > Revert.

akempkens commented:
> > ElementsController task now protected instead of private --> not sure about this (!), should more vars be protected instead
> > of private since the parent class holds protected vars and not private ones

Actually as $task is now defined in the controller it is not needed anymore within JoomFish controllers at all. This was a fallback solution for old 1.5 code together with JoomFish 2.2 being more PHP 5.3 compatible. We can remove these definitions and it should work still.

akempkens commented:
Some more feedback.

If you change the logic of files such as the language files of admin being moved it is best to do this consequently over the whole system as this allows for better maintenance. Therefore I moved all the site/language files as well.

I also learned that it is better for merging and workflow to have smaller tickets. This is why I will integrate your changes separately in different commits. This is clearly a different way to work with these systems than in SVN.

quodo commented:
Thanks for giving feedback, I learn from your comments.

I guess that future pull requests should contain one topic (instead of four here). And something like "changed mosPageNav to JPagination" is something for a pull request, but for the others it might be better to just discuss it on the forum.

klas commented:
Except changing property visibility this look like a good changes to me. 

Difference between private / protected is that children/parent can't access private ones - and since a decision between the 2 needs to be made on purpose, if this is final class (not to be extended) then private is fine, otherwise use protected. Not sure about this one but: I think declaring it private in this case would actually result in having 2 different $task properties existing in this class and the parent -and also throw strict notices as child cannot have less visibility on the same property. Sorry for spaming :) Yes, in this case it needs to be removed :)

akempkens commented:
As you say Klas there will be two different $task properties what is not our goal, so removing the variable is the way to go. In order to perform this change I need a test that proofs that everything is working correctly afterwards, this is my only stopper atm.

quodo commented:
Klas, indeed the parent class may not be stricter than the child class, that's why I got a Fatal Error. And Alex had the best idea: if the parent already holds the variable, the child doesn't need it. (Warning: my PHP programming skills are limited :-( )

Alex wrote:

> In order to perform this change I need a test that proofs that everything is working 
> correctly afterwards, this is my only stopper atm.

If I can test anything... You can easily fork the project, do the changes you want and then let other people know to test it. And after testing pull the changes into the main project.
(Proper testing should perhaps include being able to save a translation, see my note here: http://www.joomfish.net/forum/viewtopic.php?f=48&t=10770#p36964).

For example: on https://github.com/quodo/jf-future you can download your JoomFish with my changes using the Zip button: it downloads the repository as a zip file.

GeraintEdwards commented:
I'm happy with these changes.

If someone want to test some code that actually does translations you can take a look at my fork.  It provides for translation of content, articles, menus, modules using Joomla native storage and a mapping table for translations (so ULTRA fast).  The delay has been on handling the translation of JForm elements !!!

I have translations of content attributes, menu parameters and module parameters (+ menu assignments) working but the code needs quite a bit of tidying up (so no criticisms on that score thanks;) ).

Other elements can be translated using the old Joomfish approach.

akempkens commented:
Ok I will integrate the changes to both jf-core and jf-future branch, but in separate commits. As soon as I'm done I will close this request.

GeraintEdwards committed:
Merge pull request #1 from klas/master

Auto add frontent languages to available languages list

GeraintEdwards committed:
Clean out of unused files
Rename ContentObject class to TranslationObject
Start creation of custom TranslationObject classes to handle special cases

quodo commented:
I tried GeraintEdwards' copy from https://github.com/GeraintEdwards/jf-future (sun dec. 4, 2011), and saved the files to my local harddrive first. Then installed everyting "from directory" without creating a zip first.

First the good news: 
With some smaller changes (indicated below) I can use JoomFish on Joomla 1.7 an translate menu-items, modules, content, etc.! That's really great! 

And other positive feedback:
- Module mod_jflanguageselection installs without problems
- Plugin jfdatabase installs without problems
- Third pary content element files (that I tried) install fine

Not okey (yet, but solvable):

A - Installing component 'from directory'
Error installing: "Failed loading XML file"
Solution: remove /component/admin/manifest.xml (nearly empty file).
No further installation errors on the component.

B - Admin JoomFish Menu items are not translated/translatable: they use the language strings
Cause and solution: sys.ini file should be in /admin/language/en-GB/en-GB.com_joomfish.sys.ini (missing directory en-GB her), and .sys.ini file should be mentioned in manifest.xml (instead of .menu.ini file). Manifest.xml should be changed to reflect the move of the admin language files.

C - Plugin jfrouter: small error upon installation
Error: JInstaller: :Install: File does not exist \jfrouter\en-GB.plg_system_jfrouter.ini
A solution: change this in the manifest.xml file: added jfrouter/language/ directory: <language tag="en-GB">jfrouter/language/en-GB.plg_system_jfrouter.ini</language>

D - Cannot save translation:
Fatal error: Call to protected method JRequest::_stripslashesrecursive() from context 'TranslationObject' in ...\administrator\components\com_joomfish\models\TranslationObject.php on line 180
Temporary solution: comment line 180 and 181: the JRequest::_stripSlashesRecursive functions are the problem.
==> Yes, translation (of module title and tested parameters of Login form: pre- and post-text) can be saved now!
As well as content with its parameters (indeed: a new look).

E - Frontend: I wanted to view the translated module title:
Fatal error: Using $this when not in object context in ...\components\com_joomfish\helpers\joomfish.class.php on line 738
Solution: don't use $this->quote, but use $db->quote.

F - Tried "Translated powered by Google" (nice feature!)
Shows alert box: "Please use Translate v2. See http://code.google.com/apis/language/translate/overview.html" --> API v1 is no longer available per December 1, 2011.

G - Translation state is alway "up to date" (green check mark) even when there is no translation. I guess this also result in never showing a "tranlation not available" text in the frontend, setting different parameters won't change that.
(Note: for 3rd party Content Element that I tried the State is correct: Translation does not exist (red ball) when it doesn't exist, etc.)

H - Orphans menu item and Content Elements item: fatal error due to legacy pagination:
Fatal error: require_once() [function.require]: Failed opening required '.../administrator/includes/pageNavigation.php' (include_path='.;...\xampp\php\pear\;...\xampp\php\data\pear') in  ...\administrator\components\com_joomfish\controllers\translate.php on line 521
Change both occurences of 
      require_once( JPATH_SITE . "/administrator/includes/pageNavigation.php");
      $pageNav = new mosPageNav($total, $limitstart, $limit);
to
      jimport('joomla.html.pagination');
      $pageNav = new JPagination($total, $limitstart, $limit);
Files
- \admin\controllers\translate.php
- \admin\controllers\elements.php

I - JoomFish menu item: Content Elements 
Fatal error: Access level to ElementsController::$task must be protected (as in class JController) or weaker in ...\administrator\components\com_joomfish\controllers\elements.php on line 44
New solution: removed $task declaration since parent class already has it.

These were just some tests where I tried the features of JoomFish that I'm used to using. I think that although JoomFish is perhaps not ready yet to release for Joomla 1.7, I can use this (e.g. Geraints branch) already. I'm available for further testing if you'd like that :-)

GeraintEdwards commented:
Excellent feedback - great to have someone starting to test this.  Probably better as a comment on my fork directly ;)

Also many of your fixes could easily be accomodated in a pull request on my fork (I'm just learning this stuff too).

A. Correct - this file should not be there
B. Backend menu items are translated in a different way by Joomla - I didn't even try to look at these.  They should be filtered out of the list of menu items - will sort this out.
C. Correct
D. Didn't spot this because I don't have magic quotes gpc on my server. In Joomla 1.7 this test is not needed since the slashes are stripped automatically so 
E. Already fixed in the latest version
F. Klas is looking at this.
G. Not looked at this yet - hold this thought and keep reminding me ;)
H & I - not looked at these features.  I know orphans will not work - the data is stored differently.  Content elements may only need minor code tweaks.

thanks and keep testing - more feedback welcome

quodo commented:
GeraintEdwards wrote

> Probably better as a comment on my fork directly ;)
> Also many of your fixes could easily be accomodated in a pull request on my fork (I'm just learning this stuff too).

Can't seem to figure out how to 'just' comment on a fork? But pull requests I can do (and they shouldn't be like this one: too many things at once; one pull request for A and another for C would be better, I guess). 

Brings me to two questions:
1) Where should I post comments/remarks/questions/feedback? 
2) On which branch should pull requests be done? I imagine that https://github.com/JoomFish/jf-future is the master where all approved changes go?

Watch out, I'll see if and when I can get you more feedback ;-)

akempkens commented:
You can comment on the fork as you comment here. You just go to Geraint's github project and comment there. It should be possible.

To your questions:
1) We have a general board for discussions on the direction of JoomFish and its releases in our technical discussion forum: http://www.joomfish.net/forum/viewforum.php?f=15 - I added you to the tech talk group

Direct feedback to a branch I would do like here and as you did already.

2) Specific pull requests to a certain development branch can go there until this branch is planed for merging back to the master here at jf-future. Geraint's branch will be merged back as well. So this is perfectly fine and git manage that for us.

GeraintEdwards committed:
"Merge" in most of quodo's bug reports from pull request comment

quodo committed:
Changed legacy pagination:

Removed $task in class ElementsController (extends JController) since it's declared in JController

GeraintEdwards committed:
fix for various $list variables not defined in views

quodo committed:
Missing (un)published icons at the bottom of Translate/Orphans views and Control Panels Unpublished Content Elements tab. - II

akempkens commented:
Can we create new pull request for the other changes pls. It get's a bit confusing atm to follow all the different little things.

I'm closing this one to ensure we fix the right things. The work from Geraint is merged into the core jf-future branch recently.

quodo commented:
All items mentioned in this pull request are either fixed or in the issue tracker.
